src/service/action-impl.js

```diff
@@ -512,6 +531,62 @@ export class ActionService {
     });
   }
 
+  /**
+   * Overrides the node actionMap key reference based on the data provided in
+   * the metadata.
+   * @param {string} actionEventType
+   * @param {!Array<!ActionInfoDef>} actionInfos
+   * @param {?ActionEventDef} event
+   * @param {?ActionInfoMetadata=} opt_metadata
+   */
+  setNodeActionReference_(actionEventType, actionInfos, event, opt_metadata) {
+    if (!opt_metadata) {
+      return;
+    }
+    // Element that triggered the action. e.g. button.
+    const triggerNode = event.target;
+    // If the action is to override an existing action cached on a node
+    // then set it on the triggerNode and delete it on the previously set node.
+    const nodeActionCache = opt_metadata && opt_metadata['nodeActionCache'];
+    if (nodeActionCache) {
+      const actionMap = dict({});
+      const actionMapKey = nodeActionCache['action-map'];
+      actionMap[`${actionEventType}`] = actionInfos;
+      triggerNode[`${actionMapKey}`] = actionMap;
+      delete nodeActionCache[`${actionMapKey}`];
+      delete nodeActionCache['action-map'];
+    }
+  }
+
+  /**
+   * Initialize the action's arguments.
+   * @param {!ActionInfoDef} actionInfo
+   * @param {?ActionEventDef} event
+   * @param {?ActionInfoMetadata=} opt_metadata
+   * @return {?JsonObject}
+   */
+  initializeArgs_(actionInfo, event, opt_metadata) {
+    let overriddenArgs;
+    // Use arguments provided by the metadata if provided.
+    if (opt_metadata) {
+      for (const arg in actionInfo.args) {
+        const variableArgName = actionInfo.args[`${arg}`];
+        const argsAliases = /** @type {!Array} */ (opt_metadata['argsAliases']);
+        // If the arg name used in a call is a listed argument alias/expression.
+        if (argsAliases && argsAliases.includes(variableArgName)) {
+          // Then it is an expression to be evaluated against the data provided
+          // in opt_args.
+          actionInfo.args[`${arg}`] = {expression: `event.${variableArgName}`};
+        }
+      }
+      overriddenArgs = opt_metadata['args'];
+    }
```

Comment:
I wonder if we can remove this and do a simple check in amp-action-macro.js -- make sure every argument in `arguments` attribute exists as a key in `args`.

Reply:
I just checked and I don't think that unquoted strings are force-converted to `TokenType.LITERAL`. Looks like it only happens for numbers and booleans (search for "convertValues" in action-impl.js.).

Reply:
Correct but it is coerced into a string:

https://github.com/ampproject/amphtml/blob/master/src/service/action-impl.js#L976

e.g.
<button on="tap:scrollable-carousel.goToSlide(index=x)">Go to slide 1</button>

Here 'x' is initially of type ID but since the value is missing '.' delimiters to evaluate it as expression. 

But I think as you've mentioned before offline, rather than creating the expression in initializeArgs we can just tweak this part of the code to evaluate it to an expression, if the value is not a number or boolean.